pkg/storegateway/batch_series_test.go

```diff
@@ -771,52 +773,169 @@ func TestBlockSeriesChunkRefsSetsIterator_ErrorPropagation(t *testing.T) {
+				}},
+				{series: []seriesChunkRefs{
+					{lset: labels.FromStrings("a", "1", "b", "2"), chunks: make([]seriesChunkRef, 1)},
+				}},
+			},
+		},
 	}
-	suite.cache.SwapWith(noopCache{})
-
-	indexReader := firstBlock.indexReader()
-	defer indexReader.Close()
-
-	matcher, err := labels.NewMatcher(labels.MatchRegexp, "a", ".+")
-	require.NoError(t, err)
-
-	iterator, err := openBlockSeriesChunkRefsSetsIterator(
-		ctx,
-		100,
-		indexReader,
-		firstBlock.meta.ULID,
-		[]*labels.Matcher{matcher},
-		nil,
-		suite.store.seriesHashCache.GetBlockCache(firstBlock.meta.ULID.String()),
-		&limiter{limit: 1},
-		&limiter{limit: 100000},
-		false,
-		firstBlock.meta.MinTime,
-		firstBlock.meta.MaxTime,
-		nil,
-		newSafeQueryStats(),
-		suite.logger,
-	)
-	require.NoError(t, err)
 
-	_ = readAllSeriesChunkRefsSet(iterator)
-	assert.ErrorContains(t, iterator.Err(), "test limit exceeded")
+	for testName, testCase := range testCases {
+		testName, testCase := testName, testCase
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+
+			suite := prepareStoreWithTestBlocks(t, t.TempDir(), objstore.NewInMemBucket(), false, NewChunksLimiterFactory(0), NewSeriesLimiterFactory(0))
+			var firstBlock *bucketBlock
+			// Find the block with the smallest timestamp in its ULID
+			for _, b := range suite.store.blocks {
+				if firstBlock == nil {
+					firstBlock = b
+					continue
+				}
+				if b.meta.ULID.Time() < firstBlock.meta.ULID.Time() {
+					firstBlock = b
+				}
+			}
+			suite.cache.SwapWith(noopCache{})
+
+			indexReader := firstBlock.indexReader()
+			defer indexReader.Close()
+
+			iterator, err := openBlockSeriesChunkRefsSetsIterator(
+				ctx,
+				testCase.batchSize,
+				indexReader,
+				firstBlock.meta.ULID,
+				[]*labels.Matcher{testCase.matcher},
+				nil,
+				hashcache.NewSeriesHashCache(1024*1024).GetBlockCache(firstBlock.meta.ULID.String()),
+				&limiter{limit: testCase.chunksLimit},
+				&limiter{limit: testCase.seriesLimit},
+				false,
+				firstBlock.meta.MinTime,
+				firstBlock.meta.MaxTime,
+				nil,
+				newSafeQueryStats(),
+				log.NewNopLogger(),
+			)
+			require.NoError(t, err)
+
+			actualSeriesSets := readAllSeriesChunkRefsSet(iterator)
+
+			require.Lenf(t, actualSeriesSets, len(testCase.expectedSeries), "expected %d sets, but got %d", len(testCase.expectedSeries), len(actualSeriesSets))
+			for i, actualSeriesSet := range actualSeriesSets {
+				expectedSeriesSet := testCase.expectedSeries[i]
+				require.Equal(t, expectedSeriesSet.len(), actualSeriesSet.len())
+				for j, actualSeries := range actualSeriesSet.series {
+					expectedSeries := testCase.expectedSeries[i].series[j]
+
+					actualLset := actualSeries.lset
+					expectedLset := expectedSeries.lset
+					assert.Truef(t, labels.Equal(actualLset, expectedLset), "%d, %d: expected labels %s got labels %s", i, j, expectedLset, actualLset)
+
+					assert.Len(t, actualSeries.chunks, len(expectedSeries.chunks))
```

Comment:
It should be always 1 for all tests case.

```suggestion
					assert.Len(t, actualSeries.chunks, 1)
```